shared-bindings/audiodelays/Echo.c

```diff
@@ -0,0 +1,245 @@
+// This file is part of the CircuitPython project: https://circuitpython.org
+//
+// SPDX-FileCopyrightText: Copyright (c) 2024 Mark Komus
+//
+// SPDX-License-Identifier: MIT
+
+#include <stdint.h>
+
+#include "shared-bindings/audiodelays/Echo.h"
+#include "shared-module/audiodelays/Echo.h"
+
+#include "shared/runtime/context_manager_helpers.h"
+#include "py/binary.h"
+#include "py/objproperty.h"
+#include "py/runtime.h"
+#include "shared-bindings/util.h"
+
+#define DECAY_DEFAULT 0.7f
+
+//| class Echo:
+//|     """An Echo effect"""
+//|
+static mp_obj_t audiodelays_echo_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum { ARG_delay_ms, ARG_decay, ARG_buffer_size, ARG_sample_rate, ARG_bits_per_sample, ARG_samples_signed, ARG_channel_count, };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_delay_ms, MP_ARG_INT | MP_ARG_KW_ONLY, {.u_int = 50 } },
+        { MP_QSTR_decay, MP_ARG_OBJ | MP_ARG_KW_ONLY,  {.u_obj = MP_OBJ_NULL} },
```

Comment:
I suspect you'll want delay and decay to be BlockInputs so they can be varied.

https://github.com/adafruit/circuitpython/blob/87f7bc9fcc74e8ada4f255a8d7824f46a039995c/shared-bindings/synthio/__init__.c#L75-L78

Reply:
I think you could adjust your initial allocation to be larger if given a non-fixed value. That way you don't need to re-allocate when the value changes. You just don't use all of the buffer for short delays.

I also think it's good to show how to use BlockInput so that further effects use it too. It is more complicated, but it'd be good to show how to use. It is super powerful IMO.

Reply:
Delay, Mix and Decay are all BlockInputs now. I'll resolve this change and if an issue is with the BlockInputs is found a new change can be raised.